File: backend/app/domain/sinner.py
from sqlmodel import SQLModel, Field
from typing import Optional


class SinnerSnapshot(SQLModel):
    entity_index: int
    x: float
    y: float
    z: float
    spawn_time_s: int
    max_health: int
    death_time_s: Optional[int] = None
    time_alive_s: Optional[int] = None
    killer_player_slot: Optional[int] = None
    # Serde serializes HashMap<u32, i32> keys as strings, so JSON keys arrive as strings.
    # dict[str, int] matches the wire format: {"0": 160, "4": 80}
    retaliation_damage: dict[str, int] = Field(default_factory=dict)

File: frontend/src/domain/sinner.ts
export interface SinnerSnapshot {
  entity_index: number;
  x: number;
  y: number;
  z: number;
  spawn_time_s: number;
  max_health: number;
  death_time_s: number | null;
  time_alive_s: number | null;
  killer_player_slot: number | null;
  // Backend serializes Rust HashMap<u32, i32> keys as strings
  retaliation_damage: Record<string, number>;
}

export interface ScaledSinnerSnapshot extends SinnerSnapshot {
  left: number;
  top: number;
}
